File: leetcode/200. Number of Islands.py
class Solution:
    def numIslands(self, grid):
        row_count = len(grid)
        col_count = len(grid[0])

        land_count = 0

        for r in range(row_count):
            for c in range(col_count):
                if self.findLand(grid,r,c): land_count += 1

        return land_count

    def findLand(self, grid, r, c):
        if not (r >= 0 and r < len(grid)): return False
        if not (c >= 0 and c < len(grid[0])): return False
        if grid[r][c] != "1": return False

        grid[r][c] = '#'

        # expand
        self.findLand(grid,r + 1,c)
        self.findLand(grid,r - 1,c)
        self.findLand(grid,r,c + 1)
        self.findLand(grid,r,c - 1)

        return True

    
        

File: leetcode/200. Number of Islands.ts
function numIslands(grid: string[][]): number {
    let count = 0;
    for(let r = 0; r < grid.length; r++){
        for(let c = 0; c < grid[0].length; c++){
            if(grid[r][c] === '1'){
                dfs(grid,r,c)
                count++;
            }
        };
    };
    return count
};

function dfs(grid,r,c){
    if(r < 0 || c < 0 || r >= grid.length || c >= grid[0].length || grid[r][c] === '0') return
    grid[r][c] = '0'
    dfs(grid,r + 1, c)
    dfs(grid,r - 1, c)
    dfs(grid,r, c + 1)
    dfs(grid,r, c - 1)
}